File: 705_Design_HashSet.py
# Time Complexity :
#   add:    O(1)
#   pop:    O(1)
#   top:    O(1)
#   empty:  O(1)
# 
# Space Complexity : O(n)[n is the number of elements saved]
# 
# Did this code successfully run on Leetcode : Yes
# 
# Any problem you faced while coding this : No
# 
# 1. Double hashing(Mod and Division operator) has been used to avoid collisions
# 
# 
class MyHashSet:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.buckets = 1000 # The n range is 1 - 1000000. So we take the square root of upper bound as the size of the list
        self.bucketItems = 1001
        self.storage = [None for i in range(self.buckets)]
        
    def __getBucket(self, key):
        """
        Utitlity method to get the bucket index from key
        """
        return key % self.buckets
    
    def __getBucketItem(self, key):
        """
        Utitlity method to get the bucket Item index from key
        """
        return key // self.buckets
        

    def add(self, key: int) -> None:
        bucket = self.__getBucket(key)
        bucketItem = self.__getBucketItem(key)
        
        if not self.storage[bucket]: # True, when there is no element present at this bucket index
            self.storage[bucket] = [False for i in range(self.bucketItems)]
        self.storage[bucket][bucketItem] = True
        

    def remove(self, key: int) -> None:
        bucket = self.__getBucket(key)
        bucketItem = self.__getBucketItem(key)
        if self.storage[bucket]:
            self.storage[bucket][bucketItem] = False
        

    def contains(self, key: int) -> bool:
        """
        Returns true if this set contains the specified element
        """
        bucket = self.__getBucket(key)
        bucketItem = self.__getBucketItem(key)
        if self.storage[bucket]: # True only if bucket items list is present
            return self.storage[bucket][bucketItem]
        return False # Return false if there is no bucket item list


# Your MyHashSet object will be instantiated and called as such:
# obj = MyHashSet()
# obj.add(key)
# obj.remove(key)
# param_3 = obj.contains(key)

File: _705_DesignHashSet.java
import java.util.LinkedList;

// Time Complexity : o(1) for all operations
// Space Complexity : o(HASH_LEN1*HASH_LEN2+1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Approach : Double hashing collision resolution strategy.
class MyHashSet {

    final private  int HASH_LEN1 = 1000;
    final private  int HASH_LEN2 = 1000;

    private  boolean[][] hashValMatrix;

    public MyHashSet() {
        hashValMatrix = new boolean[HASH_LEN1][];
    }

    public void add(int key) {

        if(contains(key)) return;

        int hash1 = getHash1(key);
        if(hashValMatrix[hash1] == null){
            if(hash1 == 0){
               hashValMatrix[hash1] = new boolean[HASH_LEN2+1];
            }
            else{
               hashValMatrix[hash1] = new boolean[HASH_LEN2];
            }
        }
        int hash2 = getHash2(key);
        hashValMatrix[hash1][hash2] = true;

    }

    public void remove(int key) {
        if(contains(key)){
            int hash1 = getHash1(key);
            int hash2 = getHash2(key);
            hashValMatrix[hash1][hash2] = false;
        }
    }

    public boolean contains(int key) {
        int hash1 = getHash1(key);
        int hash2 = getHash2(key);
        if(hashValMatrix[hash1] == null) return false;
        return hashValMatrix[hash1][hash2];
    }

    public int getHash1(int key){
        return key%HASH_LEN1;
    }
    public int getHash2(int key){
        return key/HASH_LEN2;
    }


}
